lib/loadbalance/loadbalancer.go

```diff
@@ -0,0 +1,112 @@
+// Package loadbalance implements a least-connections load balancer.
+package loadbalance
+
+import (
+	"context"
+	"io"
+	"net"
+	"time"
+
+	"github.com/KiaFarhang/tcp-load-balancer/internal/atomic"
+)
+
+const (
+	internalServerErrorMessage          string        = "Internal server error"
+	connectionToUpstreamTimedOutMessage string        = "Timed out connecting to upstream"
+	maxConnectionTimeout                time.Duration = 3 * time.Second
+)
+
+// host is an individual server that can take traffic from the load balancer.
+type host struct {
+	address         *net.TCPAddr
+	connectionCount *atomic.Counter
+}
+
+/*
+LoadBalancer is a least-connections load balancer. It keeps track of the number
+of connections to a group of hosts, and routes a request to whichever host has the fewest
+at the time the request is processed.
+
+If two hosts have the same number of connections, the LoadBalancer will always select whichever
+host had the lower index in the list of hosts originally passed to it.
+*/
+type LoadBalancer struct {
+	hosts  []*host
+	dialer *net.Dialer
+}
+
+/*
+NewLoadBalancer constructs a new least-connections load balancer to route
+requests to the slice of TCP addresses provided. Clients should construct a separate
+LoadBalancer for each upstream application they wish to load balance.
+*/
+func NewLoadBalancer(addresses []*net.TCPAddr) *LoadBalancer {
+	var hosts []*host
+	for _, address := range addresses {
+		host := &host{address: address, connectionCount: &atomic.Counter{}}
+		hosts = append(hosts, host)
+	}
+
+	// I believe when both a context and a dialer have a timeout the shorter value
+	// is respected; this protects us from clients passing in a no-timeout context
+	// and our dial deadlocking when we can't connect to the upstream.
+	return &LoadBalancer{hosts, &net.Dialer{Timeout: maxConnectionTimeout}}
+}
+
+/*
+HandleConnection takes a TCP connection, finds a suitable host to handle it,
+then connects to the host and streams data between the connection and host until
+both sides of the connection are closed.
+
+If the connection to the host fails, HandleConnection will write an error message
+to the incoming net.Conn and close it. It will also close the net.Conn if the communication
+with the host succeeds; callers of HandleConnection do not need to close the net.Conn
+they pass in.
+*/
+func (lb *LoadBalancer) HandleConnection(ctx context.Context, conn net.Conn) {
+	host := lb.findHostWithLeastConnections()
+	connectionToHost, err := lb.dialer.DialContext(ctx, "tcp", host.address.String())
+
+	if err != nil {
+		select {
+		case <-ctx.Done():
+			conn.Write([]byte(connectionToUpstreamTimedOutMessage))
+		default:
+			conn.Write([]byte(internalServerErrorMessage))
+		}
+		conn.Close()
+		return
+	}
```

Comment:
Do you want to write the error to the connection? Since this is an arbitrary L4 LB, the other side doesn't know what to do with this error message.

What do you think about logging it here instead?

Reply:
I would follow the lead of a NLB or nginx. I don't think either return an error to the client. For example, I think nginx just logs and moves on.

Reply:
Thanks - updated to just log and close the connection.